src/silo/query_engine/actions/aggregated.cpp

```diff
@@ -0,0 +1,310 @@
+      std::partial_sort(result.begin(), result.begin() + limit.value(), result.end(), cmp);
+      result.resize(limit.value());
+   } else {
+      std::sort(result.begin(), result.end(), cmp);
+   }
+}
+
+std::vector<AggregationResult> generateResult(
+   std::unordered_map<Tuple, uint32_t>& tuple_counts,
+   const Database& database
+) {
+   std::vector<AggregationResult> result;
+   result.reserve(tuple_counts.size());
+   for (auto& [tuple, count] : tuple_counts) {
+      std::map<std::string, std::variant<std::string, int32_t, double>> fields =
+         tuple.getFields(database);
+      fields["count"] = static_cast<int32_t>(count);
+      result.push_back({fields});
+   }
+   return result;
+}
+
+QueryResult aggregateWithoutGrouping(const std::vector<OperatorResult>& bitmap_filters) {
+   uint32_t count = 0;
+   for (const auto& filter : bitmap_filters) {
+      count += filter->cardinality();
+   };
+   std::map<std::string, std::variant<std::string, int32_t, double>> tuple_fields;
+   tuple_fields["count"] = static_cast<int32_t>(count);
+   return QueryResult{std::vector<AggregationResult>{{tuple_fields}}};
+}
+
+Aggregated::Aggregated(
+   std::vector<std::string> group_by_fields,
+   std::vector<std::string> order_by_fields,
+   std::optional<uint32_t> limit
+)
+    : group_by_fields(std::move(group_by_fields)),
+      order_by_fields(std::move(order_by_fields)),
+      limit(limit) {}
+
+QueryResult Aggregated::execute(
+   const Database& database,
+   std::vector<OperatorResult> bitmap_filters
+) const {
+   if (group_by_fields.empty()) {
+      return aggregateWithoutGrouping(bitmap_filters);
+   }
+   if (group_by_fields.size() == 1) {
+      const std::string group_by_field = group_by_fields[0];
+      const auto& metadata = database.database_config.getMetadata(group_by_field);
+      if (metadata.name == database.database_config.schema.primary_key) {
+         throw QueryParseException("Cannot group by primary key field: '" + group_by_field + "'");
+      }
+      // TODO(#133) optimize when equal to partition_by field
+      // TODO(#133) optimize single field groupby
+   }
+   std::vector<config::DatabaseMetadata> group_by_metadata;
+   for (const std::string& group_by_field : group_by_fields) {
+      const auto& metadata = database.database_config.getMetadata(group_by_field);
+      group_by_metadata.push_back(metadata);
+   }
+
+   std::vector<OrderByField> order_by_definition;
+   bool randomize_order = false;
+   for (const std::string& order_by_field : order_by_fields) {
+      if (order_by_field == "random()") {
+         randomize_order = true;
+         break;
+      }
+      order_by_definition.push_back(parseOrderByField(order_by_field));
+   }
+   for (const OrderByField& field : order_by_definition) {
+      if (field.name != "count" && !std::any_of(group_by_metadata.begin(), group_by_metadata.end(), [&field](const config::DatabaseMetadata& group_by_field) {
```

Comment:
Maybe extract a constant for `"count"` as this appears several times in this file? That would make the connection a lot clearer.